tests/operators/test_agnostic_merge.py

```diff
@@ -0,0 +1,177 @@
+import math
+import pathlib
+
+import pandas as pd
+from airflow.decorators import task_group
+from airflow.utils import timezone
+
+from astro import sql as aql
+from astro.dataframe import dataframe as adf
+from astro.sql.table import Table, TempTable
+from tests.operators import utils as test_utils
+
+DEFAULT_DATE = timezone.datetime(2016, 1, 1)
+
+CWD = pathlib.Path(__file__).parent
+import pytest
+
+default_args = {
+    "owner": "airflow",
+    "retries": 1,
+    "retry_delay": 0,
+}
+
+
+def merge_keys(sql_server, mode):
+    """
+    To match with their respective API's, we have a slightly different "merge_keys" value
+    when a user is using snowflake.
+    :param sql_server:
+    :return:
+    """
+    sql_name, _ = sql_server
+    keys = []
+    if mode == "single":
+        keys = ["list"]
+    if mode == "multi":
+        keys = ["list", "sell"]
+    if mode == "mixed":
+        keys = ["list", "sell"]
+
+    if sql_name == "snowflake":
+        return {k: k for k in keys}
+    else:
+        return keys
+
+
+@pytest.fixture
+def merge_parameters(request, sql_server):
+    mode = request.param
+    if mode == "single":
+        return (
+            {
+                "merge_keys": merge_keys(sql_server, mode),
+                "merge_columns": ["list"],
+                "target_columns": ["list"],
+                "conflict_strategy": "ignore",
+            },
+            mode,
+        )
+    elif mode == "multi":
+        return (
+            {
+                "merge_keys": merge_keys(sql_server, mode),
+                "merge_columns": ["list", "sell"],
+                "target_columns": ["list", "sell"],
+                "conflict_strategy": "ignore",
+            },
+            mode,
+        )
+    elif mode == "mixed":
+        return (
+            {
+                "merge_keys": merge_keys(sql_server, mode),
+                "target_columns": ["list", "sell", "taxes"],
+                "merge_columns": ["list", "sell", "age"],
+                "conflict_strategy": "update",
+            },
+            mode,
+        )
+
+
+@aql.transform
+def do_a_thing(input_table: Table):
+    return "SELECT * FROM {{input_table}}"
+
+
+@aql.run_raw_sql
+def add_constraint(table: Table, columns):
+    if table.conn_type == "sqlite":
+        return (
+            "CREATE UNIQUE INDEX unique_index ON {{table}}" + f"({','.join(columns)})"
+        )
+    elif table.conn_type == "bigquery":
+        return ""
+    return (
+        "ALTER TABLE {{table}} ADD CONSTRAINT airflow UNIQUE"
+        + f" ({','.join(columns)})"
+    )
+
+
+@adf
+def validate_results(df: pd.DataFrame, mode, sql_type):
+    # make columns lower and reverse due to snowflake defaulting to uppercase
```

Comment:
@dimberman what is the advantage of checking over columns as opposed to rows?

Reply:
@tatiana the advantage is that it's easier to track columns than rows in dataframes. The whole reason we're doing this is that unfortunately comparing dataframes with nan values is not supported.